doc/design/clustered-etcd-alert-store/design.md

```diff
@@ -0,0 +1,150 @@
+# Clustered Etcd Alert Store
+
+Type: Design Document (and PR)
+
+Date: 2019-09-20
+
+Author: David C Wang <dcwangmit01@gmail.com>
+
+## Status Quo
+
+The current Alertmanager clustering implementation will allow multiple
+instances of Alertmanager to sync recently-updated notification log entries and
+alert silences via the gossip protocol.  The syncing of alert state is not yet
+implemented.
+
+Currently, Alertmanager alert state is entirely stored in process memory.  It
+is lost upon restart of Alertmanager.  This, in conjunction with the fact that
+Alertmanager alert state is not synced with clustered peers presents several
+problems:
+
+* Alert emitters must send alerts to every single Alertmanager instance in the
+  cluster.  If some Alertmanager instances receive alerts that other instances
+  do not, their state will drift apart.
+
+  * Typically, the alert emitter is Prometheus which uses service discovery to
+    locate all Alertmanager instances, and then sends alerts to each individual
+    target.  Custom alert emitters may not have that service discovery
+    capability and must be explicity configured.
+
+* Because the alert must be sent to every Alertmanager instance, this precludes
+  a typical HA deployment pattern of having clients (i.e. alert emitters) send
+  alerts through a Load Balancer that fronts several Alertmanager instances.
+  This pattern is typical in Kubernetes.
+
+* Restarting Alertmanager instances or increasing the replicaCount for the
+  Alertmanager cluster will result in the new members having very different
+  alert states than current members.
+
+* If all of the Alertmanagers reboot at the same time, all active alerts are
+  lost.  New alerts will be created in each instance memory as alerts come in
+  through the API.  However, these alerts will have different startsAt times
+  than the prior alerts in memory.  The effect is that "status=resolved"
+  notifications will never be sent for the original set of alerts in memory.
+  Because some notification receivers consider unique instances of alerts to be
+  a combination of alert fingerpint and startsAt, these notification reciverse
+  will never receive "status=resolved" notifications.
+
+The issues above may be addressed by implementing a share persistent data store
```

Comment:
```suggestion
The issues above may be addressed by implementing a shared persistent data store
```

Reply:
Acknowledged.  If this PR has any chance of merging I'll fix it.